keras_cv/datasets/pascal_voc/segmentation.py

```diff
@@ -0,0 +1,320 @@
+
+
+def _parse_annotation_data(annotation_file_path):
+    """Parse the annotation XML file for the image.
+
+    The annotation contains the metadata, as well as the object bounding box information.
+
+    """
+    with tf.io.gfile.GFile(annotation_file_path, "r") as f:
+        root = xml.etree.ElementTree.parse(f).getroot()
+
+        size = root.find("size")
+        width = int(size.find("width").text)
+        height = int(size.find("height").text)
+
+        objects = []
+        for obj in root.findall("object"):
+            # Get object's label name.
+            label = CLASS_TO_INDEX[obj.find("name").text.lower()]
+            # Get objects' pose name.
+            pose = obj.find("pose").text.lower()
+            is_truncated = obj.find("truncated").text == "1"
+            is_difficult = obj.find("difficult").text == "1"
+            bndbox = obj.find("bndbox")
+            xmax = int(bndbox.find("xmax").text)
+            xmin = int(bndbox.find("xmin").text)
+            ymax = int(bndbox.find("ymax").text)
+            ymin = int(bndbox.find("ymin").text)
+            objects.append(
+                {
+                    "label": label,
+                    "pose": pose,
+                    "bbox": [ymin, xmin, ymax, xmax],
+                    "is_truncated": is_truncated,
+                    "is_difficult": is_difficult,
+                }
+            )
+
+        return {"width": width, "height": height, "objects": objects}
+
+
+def _get_image_ids(data_dir, split):
+    data_file_mapping = {"train": "train.txt", "eval": "val.txt", None: "trainval.txt"}
+    with tf.io.gfile.GFile(
+        os.path.join(data_dir, "ImageSets", "Segmentation", data_file_mapping[split]),
+        "r",
+    ) as f:
+        image_ids = f.read().splitlines()
+        logging.info(f"Received {len(image_ids)} images for {split} dataset.")
+        return image_ids
+
+
+def _convert_pascal_voc_segmentation_mask_encoding(
+    segmentation_file_path, use_cache=True
+):
+    """Convert the original segmentation PNG file to proper encoding for tf.io API.
+
+    The original PNG was in a 2D (without channel dimension), which will get a wrong value
+    when directly read by tf.io API. In this function, the original image will be first
+    read with PIL.image API, expand the last dimention and save back by TF API. This will
+    ensure the converted image can be properly consumed by TF API.
+
+    If a file with `xxx_converted.png` already exists, and `use_cache` is True, the
+    conversion will be skipped.
+
+    Args:
+        segmentation_file_path: string, the original PNG file path from the dataset.
+        override: boolean, whether to use the existing cached file result.
+    Returns:
+        the file path for converted PNG file, it will be suffixed with '_converted'.
+    """
+    dir_and_file_name, _ = os.path.splitext(segmentation_file_path)
+    updated_file_path = dir_and_file_name + "_updated.png"
```

Comment:
Some high-level question -- it seems this is currently being used to directly load from raw data. In this case, do we still need to write back to a "_updated.png" file, or can we simply pass the expanded image to the data pipeline?
On the other hand, once we support "produce tfrecords", this converted data will be directly written to record file, I believe?

Reply:
If we don't write back to "x_updated.png", then we need to include some quite unreadable logic to convert the data format. This also becomes a cost per invocation of load(), whereas the x_updated.png is a one time cost.

When we change to tfrecord, it will be generated from x_updated.png as well, and the x_updated.png can be removed afterwards.

Reply:
The dataset url is temp unreachable. Have you tried to open it explicit channels with:
 ```python 
 tf.io.decode_png(
    contents,
    channels=1,
    dtype=tf.dtypes.uint8
)
```

Reply:
To tanzhenyu, I think your confusion point is why can't we just convert the PNG and make it part of the dataset without saving it. The tricky part here is the dataset building logic. When creating the dataset from metadata (a python dict), it only contains the file path (not the raw png data, otherwise it will quickly OOM). The map function for dataset will use tf.io ops to read the image, and convert them to float tensor. The loading logic could only rely on TF ops, not any PIL or pure python logic, otherwise dataset will complain and require us to use tf.pyfunction, and it will have a performance impact.

The conversion function in pure tf ops format is quite unreadable, see https://github.com/tensorflow/models/blob/master/research/deeplab/utils/get_dataset_colormap.py#L317-L331. It can be just replaced by a PIL load and convert, but the PIL can not happen in the dataset.map function.

To bhack: I had several tries before going to this solution. See following example. The 150 value is not correct, since the dataset only contains 20 classes. The 150 value should be mapped to 19 (train).

```
>>> content = tf.io.read_file("2011_000277.png")
>>> png = tf.io.decode_png(content, channels=1, dtype=tf.uint8)
>>> print(png)
tf.Tensor(
[[[150]
  [150]
  [150]
  ...
  [  0]
  [  0]
  [  0]]
......

```

Reply:
Can you share this image? The web site is still down.

Thanks.

Reply:
![2011_000277](https://user-images.githubusercontent.com/5118881/196800237-b10993e4-9f7d-4b62-959a-132330d4eb92.png)

Reply:
Thanks I have found it also in a mirror.

Yes the function it was not reading the color mapping but it is quite simple. 

You need to read it 3 channel not 1 then this is the mapping on the 20 Voc classes:

```python

VOC_COLORMAP = [[0, 0, 0], [128, 0, 0], [0, 128, 0], [128, 128, 0],
                [0, 0, 128], [128, 0, 128], [0, 128, 128], [128, 128, 128],
                [64, 0, 0], [192, 0, 0], [64, 128, 0], [192, 128, 0],
                [64, 0, 128], [192, 0, 128], [64, 128, 128], [192, 128, 128],
                [0, 64, 0], [128, 64, 0], [0, 192, 0], [128, 192, 0],
                [0, 64, 128]]
```

```python
VOC_CLASSES = ['background', 'aeroplane', 'bicycle', 'bird', 'boat',
               'bottle', 'bus', 'car', 'cat', 'chair', 'cow',
               'diningtable', 'dog', 'horse', 'motorbike', 'person',
               'potted plant', 'sheep', 'sofa', 'train', 'tv/monitor']
```

So it is not 150 but with 3 channels:
```python
Shape (375, 500, 3)

tf.Tensor(
[[[128 192   0]
  [128 192   0]
  [128 192   0]
  ...
```

That is 19 (train).

Reply:
(As we totally ignore colormapped imges libpng FLAG in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/lib/png/png_io.cc)

But as exposed above it is quite easy to remmap it (Pytorch example):
https://d2l.ai/chapter_computer-vision/semantic-segmentation-and-dataset.html

Reply:
Note that to convert the raw tensor value from 
```
[[[128 192   0]
  [128 192   0]
  [128 192   0]
```
 to
```
[[19], [19], [19]]
```
it will need a lookup from the VOC_COLORMAP, which will probably need a https://www.tensorflow.org/api_docs/python/tf/lookup/StaticHashTable. It is computation heavy, and this will happen in the map function which is not a one time cost.

Reply:
I don't know if it is the TF faster solution have you seen the link? 
Do you think that it is faster to pass twice in the filesystem?

Reply:
The temp file write will only happen once and get cached. It also happens before creating the dataset and training, whereas the lookup/map will happen within the data input pipeline.

Reply:
Thanks for the link https://d2l.ai/chapter_computer-vision/semantic-segmentation-and-dataset.html. Let me take a closer look.

Reply:
As you like:

A) We could always use the TF lookup/map in the same section as Pillow and not use an addition dependency 
B) We could always patch to use the libpng colormap flag as in pillow https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/lib/png/png_io.cc (best effect on the ecosystem)
C) We could add pillow as a dependency and use its Image

Reply:
Thanks. I agree that we can use the TF ops to do the convert rather than rely on the extra deps of Pillow.

The only concern I have is whether convert cost should happen in dataset.map or as one time pre-process step.

So far the speed I got from local temp file with Pillow is quite fast (~2 sec for the whole dataset with 10 threads). I don't have the number of the dataset.map() yet.

I think the final question is whether we prefer:
1. Temp file with pre-converted PNGs (either with PIL or TF ops)
2. dataset.map() with TF ops only (PIL doesn't work within map())

Reply:
The best solution is to patch https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/lib/png/png_io.cc as at the end both PIL and tf.io rely on libpng. So just 1 single filesystem read and decode.

But I am realistic and I know that this level of ecosystem cooperation is quite "utopian" :smile_cat: 

Then IMHO for your 2) I have an extra question are we covered by XLA with the OPset/s that we want to use?

Reply:
Good question for XLA. Let me write a TF function and have a try.

Reply:
IIUC, 
1) `create_pascal_label_colormap` is from label index to colormap, this is used for visualization;
2) `voc_label_indices` is from colormap to label index, this is for converting input mask to label.
So all we need is have a tf version of `voc_label_indices` here to make it work?

Reply:
Tested with another image, which somehow miss from the mapping. The pixel value I got from the raw png is 
[224 224 192], which doesn't map to any of the items above.

![2011_003255](https://user-images.githubusercontent.com/5118881/196820372-0c22e005-a43a-4aab-bd6c-629ba4dd340e.png)

Reply:
Is `[224 224 192]` -> `255` ?

As I see 255 with `Image.open(f)`

Reply:
I think is 255. See https://stackoverflow.com/a/55246638

Reply:
See also a TF implementation variant at https://stackoverflow.com/a/71413414 I don't know if you have a faster impl idea but this one is XLA covered.

Reply:
Thanks. I think the all the ops will work with the XLA.

```
import os

import tensorflow as tf

LOCAL_DATA_PATH = '~/pascal_voc_2012/VOCdevkit/VOC2012/'

VOC_COLORMAP = [[0, 0, 0], [128, 0, 0], [0, 128, 0], [128, 128, 0],
                [0, 0, 128], [128, 0, 128], [0, 128, 128], [128, 128, 128],
                [64, 0, 0], [192, 0, 0], [64, 128, 0], [192, 128, 0],
                [64, 0, 128], [192, 0, 128], [64, 128, 128], [192, 128, 128],
                [0, 64, 0], [128, 64, 0], [0, 192, 0], [128, 192, 0],
                [0, 64, 128]]

@tf.function(jit_compile=True)
def convert_pascal_voc_segmentation_mask_encoding(mask_file_path):
  color_map_to_label = [0] * (256 ** 3)
  for i, colormap in enumerate(VOC_COLORMAP):
    color_map_to_label[
      (colormap[0] * 256 + colormap[1]) * 256 + colormap[2]] = i
  # There is a special mapping with [224, 224, 192] -> 255
  color_map_to_label[224 * 256 * 256 + 224 * 256 + 192] = 255
  color_map_to_label = tf.constant(color_map_to_label)

  mask = tf.io.decode_png(tf.io.read_file(mask_file_path))
  mask = tf.cast(mask, tf.int32)
  mask = mask[:, :, 0] * 256 * 256 + mask[:, :, 1] * 256 + mask[:, :, 2]
  mask = tf.expand_dims(tf.gather(color_map_to_label, mask), -1)
  return mask

mask_file_path = os.path.expanduser(LOCAL_DATA_PATH) + 'SegmentationClass/2011_003255.png'
mask = convert_pascal_voc_segmentation_mask_encoding(mask_file_path)
print(mask)
```

Reply:
The creation of color_map_to_label can be lifted out, and we don't have to create a constant for every step.

Reply:
Do you think it is faster then https://stackoverflow.com/a/71413414 ?

Reply:
NIT: if you don't want to manually fill the colormap constant [check this](https://github.com/tensorflow/models/blob/238922e98dd0e8254b5c0921b241a1f5a151782f/research/deeplab/utils/get_dataset_colormap.py#L317-L331)

Reply:
Note that don't have a registered XLA kernel for `read_file` and `decode_png` so they need to be out of the XLA section. These are the optimized HLO graph of the two implementations (please note the related fusions blocks):
https://colab.research.google.com/gist/bhack/88ddd268735208d27664ca401c51f449/voc_hlo.ipynb

Reply:
We can still use XLA in tf.data as you see will still gain some fusions.

> I think this decoding logic is fine,
We have two decoding logic if you check the cloab. We just evaluating the performance between these two alternatives.

Reply:
Updated to use the tf.gather with constant tensor value lookup.